src/main/java/org/opentripplanner/routing/spt/DominanceFunction.java

```diff
@@ -65,18 +65,34 @@ public boolean betterOrEqualAndComparable(State a, State b) {
             return false;
         }
 
-        // Are the two states arriving at a vertex from two different directions where turn restrictions apply?
-        if (a.backEdge != b.getBackEdge() && (a.backEdge instanceof StreetEdge)) {
-            if (! a.getOptions().getRoutingContext().graph.getTurnRestrictions(a.backEdge).isEmpty()) {
-                return false;
-            }
+        /*
+         * The OTP algorithm tries hard to never visit the same node twice. This is generally a good idea because it avoids
+         * useless loops in the traversal leading to way faster processing time.
+         *
+         * However there is are certain rare pathological cases where through a series of turn restrictions and/or roadworks
+         * you absolutely must visit a vertex twice if you want to produce a result. One example would be a route like this:
+         *   https://tinyurl.com/ycqux93g (Note: At the time of writing this Hindenburgstr. is closed due to roadworks.)
+         *
+         * Therefore, if we are close to the start or the end of a route we allow this.
+         *
+         * More discussion: https://github.com/opentripplanner/OpenTripPlanner/issues/3393
+         *
+         * The following variable determines how close you have to be to the start or the end point for it to be
+         * considered "close enough for a loop".
+         */
+        if (a.backEdge != b.getBackEdge()
```

Comment:
Should there be a check for car mode here?

Reply:
Well, technically turn restrictions also apply to cycling, however you can always get off an push the bike. Maybe this is a good idea.

Reply:
So I checked my test case and if I only allow the loop for cars the bike will take a route like this: https://leonard.io/polyline-visualiser/?base64=b3VxZ0h9bWN1QGdBRV1VfUJhQV1RfUBddUFzQFtTQW1ARWVAQVVHQkFIS0Y_SFNIR3hA

That's ok, I guess.

Reply:
I see your point that it may be relevant to biking too. I will leave it up to you whether to include biking or. not, as you have worked more on bike routing than me